File: python/luhn/luhn.py
import re


class Luhn:
    def __init__(self, card_num):
        self.num = card_num.replace(' ', '')[::-1]

    def valid(self):
        return bool(re.fullmatch(r'\d{2,}', self.num)) and \
            not (sum([((int(num) * 2 - 9 if int(num) * 2 > 9
                      else int(num) * 2) if index % 2 else int(num))
                     for index, num in enumerate(self.num)]) % 10)

File: c/luhn/luhn.c
#include "luhn.h"
#include "assert.h"
#include "ctype.h"
#include "string.h"

#define ASCII_0      48
#define ASCII_SPACE  32
#define MIN_LENGTH   2

bool luhn(const char *num){
  assert(num);
  int length = strlen(num); 
  int digit_length = 0, sum = 0;

  for (int index = length - 1; index > -1; index--){
    char ch = num[index];
    if (isdigit(ch)){
      int value = ch - ASCII_0;
      sum += digit_length % 2 
        ? (value * 2 > 9 ? value * 2 - 9 : value * 2) 
        : value;
      ++digit_length;
    }
    else if (ch == ASCII_SPACE) continue;
    else return false;
  }

  return !(sum % 10) && digit_length >= MIN_LENGTH;
}
